From c3bea09417ec13305fa950a23ec5dd193991a74b Mon Sep 17 00:00:00 2001
From: Zhou Guanghui <zhouguanghui1@huawei.com>
Date: Tue, 25 Oct 2022 11:58:03 +0000
Subject: [PATCH 2/4] driver/smmuv3: TTB1 is supported

--------------------------------------------

If the domain attribute is DOMAIN_ATTR_TTB1, enable ttb1 of the context
descriptor, including scenario when only ttb1(for dma remapping) is
used and scenario where both ttb1 and ttb0(for sva) are enabled.

The configuration in the context descriptor would typically look like
this.
Before:
                          tables of domain A
                           .->+--------+
                          / 0 |  ttb0  |-------> io_pgtable
                         /    |  ttb1  |
             Device     /     +--------+
           +--------+  /    1 |  ttb0  |-------> pgd X
           |      A |-'       |  ttb1  |
           +--------+         +--------+
           :        :       2 |  ttb0  |-------> pgd Y
           +--------+         |  ttb1  |
                              +--------+

After:
                          tables of domain A
                           .->+--------+
                          / 0 |  ttb0  |
                         /    |  ttb1  |-------> io_pgtable
             Device     /     +--------+    /
           +--------+  /    1 |  ttb0  |-- | --> pgd X
           |      A |-'       |  ttb1  |---+
           +--------+         +--------+   |
           :        :       2 |  ttb0  |-- | --> pgd Y
           +--------+         |  ttb1  |---'
                              +--------+

Signed-off-by: Zhou Guanghui <zhouguanghui1@huawei.com>
---
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c | 67 +++++++++++++++++----
 drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h | 12 ++++
 2 files changed, 67 insertions(+), 12 deletions(-)

diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
index 13d78f492..ca837bfb7 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.c
@@ -1256,6 +1256,7 @@ int arm_smmu_write_ctx_desc(struct arm_smmu_domain *smmu_domain, int ssid,
 	bool cd_live;
 	__le64 *cdptr;
 	struct arm_smmu_device *smmu = smmu_domain->smmu;
+	struct arm_smmu_ctx_desc *domain_cd = &smmu_domain->s1_cfg.cd;
 
 	if (WARN_ON(ssid >= (1 << smmu_domain->s1_cfg.s1cdmax)))
 		return -E2BIG;
@@ -1281,9 +1282,22 @@ int arm_smmu_write_ctx_desc(struct arm_smmu_domain *smmu_domain, int ssid,
 	} else { /* (1) and (2) */
 		u64 tcr = cd->tcr;
 
-		cdptr[1] = cpu_to_le64(cd->ttbr & CTXDESC_CD_1_TTB0_MASK);
-		cdptr[2] = 0;
-		cdptr[3] = cpu_to_le64(cd->mair);
+		if (!smmu_domain->ttb1_enabled) {
+			cdptr[1] = cpu_to_le64(cd->ttbr & CTXDESC_CD_1_TTB0_MASK);
+			cdptr[2] = 0;
+			cdptr[3] = cpu_to_le64(cd->mair);
+		} else if (ssid == 0) {
+			cdptr[1] = 0;
+			cdptr[2] = cpu_to_le64(cd->ttbr & CTXDESC_CD_2_TTB1_MASK);
+			cdptr[3] = cpu_to_le64(FIELD_PREP(CTXDESC_CD_3_MAIR1, cd->mair));
+		} else {
+			cdptr[1] = cpu_to_le64(cd->ttbr & CTXDESC_CD_1_TTB0_MASK);
+			cdptr[2] = cpu_to_le64(domain_cd->ttbr & CTXDESC_CD_2_TTB1_MASK);
+			cdptr[3] = cpu_to_le64(FIELD_PREP(CTXDESC_CD_3_MAIR0, cd->mair) |
+					FIELD_PREP(CTXDESC_CD_3_MAIR1, domain_cd->mair));
+			tcr |= domain_cd->tcr;
+			tcr &= ~(CTXDESC_CD_0_TCR_EPD0 | CTXDESC_CD_0_TCR_EPD1);
+		}
 
 		if (!(smmu->features & ARM_SMMU_FEAT_HD))
 			tcr &= ~CTXDESC_CD_0_TCR_HD;
@@ -2539,14 +2553,25 @@ static int arm_smmu_domain_finalise_cd(struct arm_smmu_domain *smmu_domain,
 
 	cfg->cd.asid	= (u16)asid;
 	cfg->cd.ttbr	= pgtbl_cfg->arm_lpae_s1_cfg.ttbr;
-	cfg->cd.tcr	= FIELD_PREP(CTXDESC_CD_0_TCR_T0SZ, tcr->tsz) |
-			  FIELD_PREP(CTXDESC_CD_0_TCR_TG0, tcr->tg) |
-			  FIELD_PREP(CTXDESC_CD_0_TCR_IRGN0, tcr->irgn) |
-			  FIELD_PREP(CTXDESC_CD_0_TCR_ORGN0, tcr->orgn) |
-			  FIELD_PREP(CTXDESC_CD_0_TCR_SH0, tcr->sh) |
-			  FIELD_PREP(CTXDESC_CD_0_TCR_IPS, tcr->ips) |
-			  CTXDESC_CD_0_TCR_HA | CTXDESC_CD_0_TCR_HD |
-			  CTXDESC_CD_0_TCR_EPD1 | CTXDESC_CD_0_AA64;
+	if (!smmu_domain->ttb1_enabled)
+		cfg->cd.tcr = FIELD_PREP(CTXDESC_CD_0_TCR_T0SZ, tcr->tsz) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_TG0, tcr->tg) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_IRGN0, tcr->irgn) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_ORGN0, tcr->orgn) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_SH0, tcr->sh) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_IPS, tcr->ips) |
+			      CTXDESC_CD_0_TCR_HA | CTXDESC_CD_0_TCR_HD |
+			      CTXDESC_CD_0_TCR_EPD1 | CTXDESC_CD_0_AA64;
+	else
+		cfg->cd.tcr = FIELD_PREP(CTXDESC_CD_0_TCR_T1SZ, tcr->tsz) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_TG1, tcr->tg) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_IRGN1, tcr->irgn) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_ORGN1, tcr->orgn) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_SH1, tcr->sh) |
+			      FIELD_PREP(CTXDESC_CD_0_TCR_IPS, tcr->ips) |
+			      CTXDESC_CD_0_TCR_HA | CTXDESC_CD_0_TCR_HD |
+			      CTXDESC_CD_0_TCR_EPD0 | CTXDESC_CD_0_AA64;
+
 	cfg->cd.mair	= pgtbl_cfg->arm_lpae_s1_cfg.mair;
 	return 0;
 }
@@ -2632,6 +2657,7 @@ static int arm_smmu_domain_finalise(struct iommu_domain *domain,
 				 struct io_pgtable_cfg *);
 	struct arm_smmu_domain *smmu_domain = to_smmu_domain(domain);
 	struct arm_smmu_device *smmu = smmu_domain->smmu;
+	bool ttb1_enabled = smmu_domain->ttb1_enabled;
 
 	if (domain->type == IOMMU_DOMAIN_IDENTITY) {
 		smmu_domain->stage = ARM_SMMU_DOMAIN_BYPASS;
@@ -2693,12 +2719,26 @@ static int arm_smmu_domain_finalise(struct iommu_domain *domain,
 	else if (smmu->features & ARM_SMMU_FEAT_BBML2)
 		pgtbl_cfg.quirks |= IO_PGTABLE_QUIRK_ARM_BBML2;
 
+	if (ttb1_enabled)
+		pgtbl_cfg.quirks |= IO_PGTABLE_QUIRK_ARM_TTBR1;
+
 	pgtbl_ops = alloc_io_pgtable_ops(fmt, &pgtbl_cfg, smmu_domain);
 	if (!pgtbl_ops)
 		return -ENOMEM;
 
 	domain->pgsize_bitmap = pgtbl_cfg.pgsize_bitmap;
-	domain->geometry.aperture_end = (1UL << pgtbl_cfg.ias) - 1;
+	if (domain->geometry.aperture_end) {
+		if (ttb1_enabled &&
+		    (~0UL << pgtbl_cfg.ias) > domain->geometry.aperture_end) {
+			free_io_pgtable_ops(pgtbl_ops);
+			return -EINVAL;
+		}
+	} else if (ttb1_enabled) {
+		domain->geometry.aperture_start = ~0UL << ias;
+		domain->geometry.aperture_end = ~0UL;
+	} else {
+		domain->geometry.aperture_end = (1UL << pgtbl_cfg.ias) - 1;
+	}
 	domain->geometry.force_aperture = true;
 
 	ret = finalise_stage_fn(smmu_domain, master, &pgtbl_cfg);
@@ -3406,6 +3446,9 @@ static int arm_smmu_domain_set_attr(struct iommu_domain *domain,
 			else
 				smmu_domain->stage = ARM_SMMU_DOMAIN_S1;
 			break;
+		case DOMAIN_ATTR_TTB1:
+			smmu_domain->ttb1_enabled = *(int *)data;
+			break;
 		default:
 			ret = -ENODEV;
 		}
diff --git a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h
index e39fd52b2..99cbb060d 100644
--- a/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h
+++ b/drivers/iommu/arm/arm-smmu-v3/arm-smmu-v3.h
@@ -326,6 +326,12 @@
 #define CTXDESC_CD_0_TCR_ORGN0		GENMASK_ULL(11, 10)
 #define CTXDESC_CD_0_TCR_SH0		GENMASK_ULL(13, 12)
 #define CTXDESC_CD_0_TCR_EPD0		(1ULL << 14)
+
+#define CTXDESC_CD_0_TCR_T1SZ		GENMASK_ULL(21, 16)
+#define CTXDESC_CD_0_TCR_TG1		GENMASK_ULL(23, 22)
+#define CTXDESC_CD_0_TCR_IRGN1		GENMASK_ULL(25, 24)
+#define CTXDESC_CD_0_TCR_ORGN1		GENMASK_ULL(27, 26)
+#define CTXDESC_CD_0_TCR_SH1		GENMASK_ULL(29, 28)
 #define CTXDESC_CD_0_TCR_EPD1		(1ULL << 30)
 
 #define CTXDESC_CD_0_ENDI		(1UL << 15)
@@ -333,6 +339,7 @@
 
 #define CTXDESC_CD_0_TCR_IPS		GENMASK_ULL(34, 32)
 #define CTXDESC_CD_0_TCR_TBI0		(1ULL << 38)
+#define CTXDESC_CD_0_TCR_TBI1		(1ULL << 39)
 
 #define CTXDESC_CD_0_TCR_HA		(1UL << 43)
 #define CTXDESC_CD_0_TCR_HD		(1UL << 42)
@@ -345,6 +352,10 @@
 #define CTXDESC_CD_0_ASID		GENMASK_ULL(63, 48)
 
 #define CTXDESC_CD_1_TTB0_MASK		GENMASK_ULL(51, 4)
+#define CTXDESC_CD_2_TTB1_MASK		GENMASK_ULL(51, 4)
+
+#define CTXDESC_CD_3_MAIR0		GENMASK_ULL(31, 0)
+#define CTXDESC_CD_3_MAIR1		GENMASK_ULL(63, 32)
 
 #define CTXDESC_CD_5_PARTID_MASK	GENMASK_ULL(47, 32)
 #define CTXDESC_CD_5_PMG_MASK		GENMASK_ULL(55, 48)
@@ -810,6 +821,7 @@ struct arm_smmu_domain {
 
 	struct io_pgtable_ops		*pgtbl_ops;
 	bool				non_strict;
+	bool				ttb1_enabled;
 	bool				stall_enabled;
 	atomic_t			nr_ats_masters;
 
-- 
2.17.1

